lib/FrontendTool/ImportedModules.cpp

```diff
@@ -11,43 +11,99 @@
 //===----------------------------------------------------------------------===//
 
 #include "ImportedModules.h"
+#include "swift/AST/ASTContext.h"
 #include "swift/AST/Decl.h"
 #include "swift/AST/DiagnosticEngine.h"
 #include "swift/AST/DiagnosticsFrontend.h"
 #include "swift/AST/Module.h"
+#include "swift/Basic/STLExtras.h"
+#include "swift/ClangImporter/ClangImporter.h"
 #include "swift/Frontend/FrontendOptions.h"
+#include "clang/Basic/Module.h"
 #include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/StringRef.h"
 #include "llvm/Support/FileSystem.h"
 
-bool swift::emitImportedModules(DiagnosticEngine &diags, ModuleDecl *module,
+using llvm::StringRef;
+
+static StringRef getTopLevelName(const clang::Module *module) {
+  return module->getTopLevelModule()->Name;
+}
+
+static void findAllClangImports(const clang::Module *module,
+                                llvm::SetVector<StringRef> &modules) {
+  for (auto imported : module->Imports) {
+    modules.insert(getTopLevelName(imported));
+  }
+
+  for (auto sub : swift::makeIteratorRange(module->submodule_begin(),
+                                           module->submodule_end())) {
+    findAllClangImports(sub, modules);
+  }
+}
+
+bool swift::emitImportedModules(ASTContext &Context, ModuleDecl *mainModule,
                                 const FrontendOptions &opts) {
 
   auto path = opts.getSingleOutputFilename();
   std::error_code EC;
   llvm::raw_fd_ostream out(path, EC, llvm::sys::fs::F_None);
 
   if (out.has_error() || EC) {
-    diags.diagnose(SourceLoc(), diag::error_opening_output, path, EC.message());
+    Context.Diags.diagnose(SourceLoc(), diag::error_opening_output, path,
+                           EC.message());
     out.clear_error();
     return true;
   }
 
-  llvm::SmallVector<Decl *, 32> Decls;
-  module->getDisplayDecls(Decls);
+  llvm::SetVector<StringRef> Modules;
 
-  llvm::SetVector<Identifier> Modules;
-  for (auto *D : Decls) {
+  // Find the imports in the main Swift code.
+  llvm::SmallVector<Decl *, 32> Decls;
+  mainModule->getDisplayDecls(Decls);
+  for (auto D : Decls) {
     auto ID = dyn_cast<ImportDecl>(D);
     if (!ID)
       continue;
 
     auto accessPath = ID->getModulePath();
     // only the top-level name is needed (i.e. A in A.B.C)
-    Modules.insert(accessPath[0].first);
+    Modules.insert(accessPath[0].first.str());
+  }
+
+  // And now look in the C code we're possibly using.
+  auto clangImporter =
+      static_cast<ClangImporter *>(Context.getClangModuleLoader());
```

Comment:
> The code that looks for an adapter/overlay is just going through the usual module-loading path IIRC. I would expect it to find the dummy Foundation.swiftmodule and then fail to load it.

Yes, I believe that's the failure I mentioned above.

My confusion is why this behavior matters: won't either the overlay exist and be valid (in which case loading it is perfect fine), or the overlay not exist (so we don't load it, also fine)? It seems... surprising to me if something using this was to insert invalid swiftmodules into the filesystem?